espnet2/enh/espnet_model.py

```diff
@@ -8,51 +9,84 @@
 
 
+is_torch_1_3_plus = LooseVersion(torch.__version__) >= LooseVersion("1.3.0")
+ALL_LOSS_TYPES = (
+    # mse_loss(predicted_mask, target_label)
+    "mask_mse",
+    # mse_loss(enhanced_magnitude_spectrum, target_magnitude_spectrum)
+    "magnitude",
+    # mse_loss(enhanced_complex_spectrum, target_complex_spectrum)
+    "spectrum",
+    # log_mse_loss(enhanced_complex_spectrum, target_complex_spectrum)
+    "spectrum_log",
+    # si_snr(enhanced_waveform, target_waveform)
+    "si_snr",
+)
+EPS = torch.finfo(torch.get_default_dtype()).eps
+
+
 class ESPnetEnhancementModel(AbsESPnetModel):
     """Speech enhancement or separation Frontend model"""
 
     def __init__(
         self,
-        enh_model: Optional[AbsEnhancement],
+        encoder: AbsEncoder,
+        separator: AbsSeparator,
+        decoder: AbsDecoder,
+        stft_consistency: bool = False,
+        loss_type: str = "mask_mse",
+        mask_type: Optional[str] = None,
     ):
         assert check_argument_types()
 
         super().__init__()
 
-        self.enh_model = enh_model
-        self.num_spk = enh_model.num_spk
-        self.num_noise_type = getattr(self.enh_model, "num_noise_type", 1)
-        # get mask type for TF-domain models
-        self.mask_type = getattr(self.enh_model, "mask_type", None)
+        self.encoder = encoder
+        self.separator = separator
+        self.decoder = decoder
+        self.num_spk = separator.num_spk
+        self.num_noise_type = getattr(self.separator, "num_noise_type", 1)
+
+        if loss_type != "si_snr" and isinstance(encoder, ConvEncoder):
+            raise TypeError(f"{loss_type} is not supported with {type(ConvEncoder)}")
+
+        # get mask type for TF-domain models (only used when loss_type="mask_*")
+        self.mask_type = mask_type.upper() if mask_type else None
         # get loss type for model training
-        self.loss_type = getattr(self.enh_model, "loss_type", None)
-        assert self.loss_type in (
-            # mse_loss(predicted_mask, target_label)
-            "mask_mse",
-            # mse_loss(enhanced_magnitude_spectrum, target_magnitude_spectrum)
-            "magnitude",
-            # mse_loss(enhanced_complex_spectrum, target_complex_spectrum)
-            "spectrum",
-            # si_snr(enhanced_waveform, target_waveform)
-            "si_snr",
-        ), self.loss_type
+        self.loss_type = loss_type
+        # whether to compute the TF-domain loss while enforcing STFT consistency
+        self.stft_consistency = stft_consistency
+
+        if stft_consistency and loss_type in ["mask_mse", "si_snr"]:
+            raise ValueError(
+                f"stft_consistency will not work when '{loss_type}'' loss is used"
+            )
+
+        assert self.loss_type in ALL_LOSS_TYPES, self.loss_type
         # for multi-channel signal
-        self.ref_channel = getattr(self.enh_model, "ref_channel", -1)
+        self.ref_channel = getattr(self.separator, "ref_channel", -1)
```

Comment:
Better to pass the argument "ref_channel" to both Separator and ESPnetEnhancementModel.